Opam pin version conflict after cleaning

opam looses pinned version if a clean is performed before install:
```
$ opam pin foo.1+dev . -n
Package foo does not exist, create as a NEW package? [Y/n] y
[foo.1+dev] synchronised from git+file:///tmp/pindev#master
foo is now pinned to git+file:///tmp/pindev#master (version 1+dev)

$ opam clean
Cleaning up switch testing
Clearing cache of downloaded files
Clearing logs

$ opam install foo.1+dev
<><> Synchronising pinned packages ><><><><><><><><><><><><><><><><><><><><><><>
[foo.1+dev] synchronised from git+file:///tmp/pindev#master
[foo] Conflicting update of the metadata from git+file:///tmp/pindev#master.
Override files in .opam/testing/.opam-switch/overlay/foo (there will be a backup)? [Y/n] y
User metadata backed up in opam/backup/foo.bak

[ERROR] foo = 1+dev not available because the package is pinned to version 1
```

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -2935,16 +2935,16 @@ let clean =
            List.iter (fun d ->
                if not (List.mem d pinning_overlay_dirs) then rmdir d)
              (OpamFilename.dirs (OpamPath.Switch.Overlay.dir root sw));
-           let installed_dev_dirs =
-             OpamPackage.Set.fold (fun nv acc ->
-                 if OpamSwitchState.is_dev_package st nv
-                 then OpamSwitchState.source_dir st nv :: acc
-                 else acc)
-               st.installed []
+           let keep_sources_dir =
+             OpamPackage.Set.elements
+               (OpamPackage.Set.union st.pinned
+                  (OpamPackage.Set.filter (OpamSwitchState.is_dev_package st)
+                     st.installed))
+             |> List.map (OpamSwitchState.source_dir st)
            in
            OpamFilename.dirs (OpamPath.Switch.sources_dir root sw) |>
            List.iter (fun d ->
-               if not (List.mem d installed_dev_dirs) then rmdir d))
+               if not (List.mem d keep_sources_dir) then rmdir d))
          switches);
     if repos then
       (OpamFilename.with_flock `Lock_write (OpamPath.repos_lock gt.root)

